src/Function.FromCallback.cs

```diff
@@ -81,6 +81,8 @@ public static Function FromCallback(IStore store, Action callback)
                     out var externFunc
                 );
 
+                GC.KeepAlive(store);
```

Comment:
Shouldn't `store` remain stack-rooted for the duration of this function given it's a parameter to this method and passed as an argument to `Function`'s constructor (which will then store the reference)? On the surface, I don't see how `store` could be GC'd at any point in this method.

I'm not in anyway familiar with the optimizations the JIT compiler can make to shorten the root's lifetime in a way that would impact the correctness of this code, though. 

Would you mind further elaborate on the need for `GC.KeepAlive` here?

Reply:
I see, perhaps this guards against `Function`'s constructor call getting inlined and the caller discards the return value (at every call site? does the JIT compiler even make that optimization?), thus optimizing away any further use of the `store` parameter such that the lifetime is shortened to just after the `store.Context.handle` access?

But if this happens, it means the caller isn't using the `store` argument after this call either and since it discarded the returned `Function` (which must keep `store` reachable), then no other store-related objects are reachable either.

If that is true, then `store` can be finalized at this `GC.KeepAlive` site without a correctness issue, no?

Reply:
Except `store.Context.handle` is `IntPtr` and not `Handle`, which isn't being rooted _during_ the call to `wasmtime_func_new`.

Thus any attempts internally in the Wasmtime API to access the `Store` might be unsafe.

Now I see the need for it! Sorry for the manual thought process here.

Reply:
Thanks! Yes, the issue is that the `Store.Handle` instance isn't rooted during the `wasmtime_func_new` call, which can mean that the GC already collects the `Function` instance together with referenced the `Store` instance and thus the Store's `Handle` instance (if the caller doesn't use the returned `Function`) and calls its finalizer, and so it could happen that `wasmtime_store_delete` has already been called or is currently being called (from the GC finalizer thread) when the current thread enters the native `wasmtime_func_new` where the store context obtained from the store is passed as `IntPtr`.

To prevent this, `GC.KeepAlive()` can used, which does nothing by itself, but instructs the runtime to treat the specified reference as being reachable until after that call. This pattern is also used by the .NET runtime libraries itself, for example in the `System.Windows.Forms.Control` class (there, it prevents the Windows `hWnd` from being freed): https://github.com/dotnet/winforms/blob/9184e02413ee442f30d2548db84088bb7336edcc/src/System.Windows.Forms/src/System/Windows/Forms/DateTimePicker.cs#L1535-L1536

Alternatively (at least when we actually have a `Store`, not an `IStore`), instead of using `GC.KeepAlive(store)` we could also use `GC.KeepAlive(store.Handle)` which would be a bit more accurate, as it's actually the `Store.Handle` instance that contains the finalizer which we need to prevent from being called.

Note that native methods that operate with the `Store` itself, like `wasmtime_store_context` (as well as other types like `Engine` etc.), are currently implemented by specifying a `Handle` instance that derives from `SafeHandle`, for which the .NET runtime uses marshalling logic to keep the `SafeHandle` instance alive during the native call. However, this requires an additional allocation for the `SafeHandle` as it's a class, and the marshalling logic might cost a bit of performance.

When optimizing for performance, it might make sense to consider replacing the `SafeHandle` with just an `IntPtr` as native method parameter, and using `GC.KeepAlive()` calls instead to keep the finalizable object alive. This would avoids the separate allocation of `SafeHandle` and would improves performance for the native call (when only using blittable types). However currently I don't see a need for this.

Thanks!